File: src/features/markdown/types/markdownAst.ts
export type SourceRange = { start: number; end: number };

export type MarkdownNode = {
  id: string;
  range: SourceRange;
} & MarkdownNodeKind;

export type MarkdownNodeKind =
  | { type: "paragraph" | "blockQuote" | "listItem" | "definitionList" | "definitionTitle" | "definition" | "tableHead" | "tableRow" | "tableCell" | "emphasis" | "strong" | "strikethrough" | "superscript" | "subscript" | "metadata"; children: MarkdownNode[] }
  | { type: "heading"; level: number; children: MarkdownNode[] }
  | { type: "codeBlock"; language?: string; children: MarkdownNode[] }
  | { type: "list"; start?: number; children: MarkdownNode[] }
  | { type: "footnoteDefinition"; label: string; children: MarkdownNode[] }
  | { type: "table"; alignments: string[]; children: MarkdownNode[] }
  | { type: "link" | "image"; url: string; title: string; children: MarkdownNode[] }
  | { type: "text" | "inlineCode" | "html"; text: string }
  | { type: "math"; text: string; display: boolean }
  | { type: "footnoteReference"; label: string }
  | { type: "taskListMarker"; checked: boolean }
  | { type: "softBreak" | "hardBreak" | "rule" };

export type MarkdownOutlineNode = {
  id: string;
  nodeId: string;
  level: number;
  title: string;
  children: MarkdownOutlineNode[];
};

export type MarkdownDiagnostic = {
  severity: string;
  message: string;
  range?: SourceRange;
};

export type StructuredMarkdownDocument = {
  documentId: string;
  sessionId: string;
  revision: number;
  path: string;
  title: string;
  source: string;
  nodes: MarkdownNode[];
  outline: MarkdownOutlineNode[];
  diagnostics: MarkdownDiagnostic[];
};

export type MarkdownEdit = { startUtf16: number; endUtf16: number; text: string };

File: src-tauri/src/domain/markdown_ast.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct SourceRange {
    pub(crate) start: usize,
    pub(crate) end: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct MarkdownDocumentModel {
    pub(crate) document_id: String,
    pub(crate) session_id: String,
    pub(crate) revision: u64,
    pub(crate) path: String,
    pub(crate) title: String,
    pub(crate) source: String,
    pub(crate) nodes: Vec<MarkdownNode>,
    pub(crate) outline: Vec<MarkdownOutlineNode>,
    pub(crate) diagnostics: Vec<MarkdownDiagnostic>,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct ParsedMarkdown {
    pub(crate) nodes: Vec<MarkdownNode>,
    pub(crate) outline: Vec<MarkdownOutlineNode>,
    pub(crate) diagnostics: Vec<MarkdownDiagnostic>,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct MarkdownNode {
    pub(crate) id: String,
    pub(crate) range: SourceRange,
    pub(crate) kind: MarkdownNodeKind,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum MarkdownNodeKind {
    Paragraph {
        children: Vec<MarkdownNode>,
    },
    Heading {
        level: u8,
        children: Vec<MarkdownNode>,
    },
    BlockQuote {
        children: Vec<MarkdownNode>,
    },
    CodeBlock {
        language: Option<String>,
        children: Vec<MarkdownNode>,
    },
    List {
        start: Option<u64>,
        children: Vec<MarkdownNode>,
    },
    ListItem {
        children: Vec<MarkdownNode>,
    },
    FootnoteDefinition {
        label: String,
        children: Vec<MarkdownNode>,
    },
    DefinitionList {
        children: Vec<MarkdownNode>,
    },
    DefinitionTitle {
        children: Vec<MarkdownNode>,
    },
    Definition {
        children: Vec<MarkdownNode>,
    },
    Table {
        alignments: Vec<String>,
        children: Vec<MarkdownNode>,
    },
    TableHead {
        children: Vec<MarkdownNode>,
    },
    TableRow {
        children: Vec<MarkdownNode>,
    },
    TableCell {
        children: Vec<MarkdownNode>,
    },
    Emphasis {
        children: Vec<MarkdownNode>,
    },
    Strong {
        children: Vec<MarkdownNode>,
    },
    Strikethrough {
        children: Vec<MarkdownNode>,
    },
    Superscript {
        children: Vec<MarkdownNode>,
    },
    Subscript {
        children: Vec<MarkdownNode>,
    },
    Link {
        url: String,
        title: String,
        children: Vec<MarkdownNode>,
    },
    Image {
        url: String,
        title: String,
        children: Vec<MarkdownNode>,
    },
    Metadata {
        children: Vec<MarkdownNode>,
    },
    Text {
        text: String,
    },
    InlineCode {
        text: String,
    },
    Html {
        text: String,
    },
    Math {
        text: String,
        display: bool,
    },
    FootnoteReference {
        label: String,
    },
    SoftBreak,
    HardBreak,
    Rule,
    TaskListMarker {
        checked: bool,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct MarkdownOutlineNode {
    pub(crate) id: String,
    pub(crate) node_id: String,
    pub(crate) level: u8,
    pub(crate) title: String,
    pub(crate) children: Vec<MarkdownOutlineNode>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct MarkdownDiagnostic {
    pub(crate) severity: String,
    pub(crate) message: String,
    pub(crate) range: Option<SourceRange>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct MarkdownEdit {
    pub(crate) start_utf16: usize,
    pub(crate) end_utf16: usize,
    pub(crate) text: String,
}
